core/src/main/java/org/elasticsearch/search/fetch/matchedqueries/MatchedQueriesFetchSubPhase.java

```diff
@@ -43,7 +43,7 @@
 
     @Override
     public void hitsExecute(SearchContext context, InternalSearchHit[] hits) {
-        if (hits.length == 0) {
+        if (hits.length == 0 || context.parsedQuery() == null) {
```

Comment:
this is not equivalent to a match all, the parsedQuery is null only when the search request only has suggest and we skip the search context preprocessing [here](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/search/internal/DefaultSearchContext.java#L192), ensuring we don't run a match all query when a search request only has suggest and no explicit search query